### PR TITLE
feat(Request.signal) Initial support for signal in Request + fetch and Request + Bun.serve

### PR DESCRIPTION
**Pull request by cirospaciari**

### CODE REVIEW

File: src/bun.js/webcore/request.zig [hunk 1]
```diff
@@ -415,6 +435,23 @@ pub const Request = struct {
...
                 }
 
+                if (arguments[1].get(globalThis, "signal")) |signal_| {
+                    if (AbortSignal.fromJS(signal_)) |signal| {
+                        //Keep it alive
+                        request.js_signal = JSC.Strong.create(
+                            signal_,
+                            globalThis,
+                        );
+                        request.signal = signal;
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/request.zig:443]: just to be sure stack scanning sees it

`signal.ensureStillAlive()`

File: src/bun.js/webcore/request.zig [hunk 2]
```diff
@@ -415,6 +435,23 @@ pub const Request = struct {
                     }
                 }
 
+                if (arguments[1].get(globalThis, "signal")) |signal_| {
+                    if (AbortSignal.fromJS(signal_)) |signal| {
+                        //Keep it alive
+                        request.js_signal = JSC.Strong.create(
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/request.zig]: reminder that this shouldn't be strong anymore

File: test/bun.js/fetch.test.js [hunk 3]
```diff
@@ -1,71 +1,220 @@
...
+  it("AbortErrorWhileStreaming", async () => {
+    const abortController = new AbortController();
+    let error;
+    try {
+      let counter = 100;
+      await fetch(
+        "http://localhost:64321/stream",
+        {
+          method: "POST",
+          body: new ReadableStream({
```
Comment (Jarred-Sumner) [inline test/bun.js/fetch.test.js]: `start` is executed synchronously, so this test does nothing

**Comment by Jarred-Sumner** [comment]
I don't think we can ship this until it also works with Bun.serve() because people will expect it to work there too

**Comment by cirospaciari** [comment]
> I don't think we can ship this until it also works with Bun.serve() because people will expect it to work there too

added

File: src/bun.js/bindings/webcore/AbortSignal.cpp [hunk 4]
```diff
@@ -115,6 +115,15 @@ void AbortSignal::signalAbort(JSC::JSValue reason)
     dispatchEvent(Event::create(eventNames().abortEvent, Event::CanBubble::No, Event::IsCancelable::No));
 }
 
+void AbortSignal::cleanNativeBindings(void* ref) {
+    auto callbacks = std::exchange(m_native_callbacks, {});
+
+    callbacks.removeLastMatching([=](auto callback){
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/webcore/AbortSignal.cpp]: should it be removeLastMatching or remove all matching?

File: src/bun.js/api/server.zig [hunk 5]
```diff
@@ -1011,6 +1011,15 @@ fn NewRequestContext(comptime ssl_enabled: bool, comptime debug_mode: bool, comp
                     // User called .blob(), .json(), text(), or .arrayBuffer() on the Request object
                     // but we received nothing or the connection was aborted
                     if (request_js.as(Request)) |req| {
+                        if (req.signal) |signal| {
+                            // if signal is not aborted, abort the signal
+                            if(!signal.aborted()){
```
Comment (Jarred-Sumner) [inline src/bun.js/api/server.zig]: There are two places we need this code in, grep for the comment above

Comment (cirospaciari) [reply src/bun.js/api/server.zig]: I see fixed on https://github.com/oven-sh/bun/pull/2097/commits/483a9f85c384626b9c42740eb7a6421d79510367

File: src/bun.js/bindings/webcore/AbortSignal.cpp (hunk 4 above)
Comment (cirospaciari) [reply src/bun.js/bindings/webcore/AbortSignal.cpp]: `removeAllMatching` is the right answer, fixed on https://github.com/oven-sh/bun/commit/483a9f85c384626b9c42740eb7a6421d79510367

**Comment by Jarred-Sumner** [comment]
can you add a test to the list of ReactDOM tests which sees how that handles aborting in the middle of a ReadableStream? 

after that I think one or 2 more tests with ReadableStream `type`: `"direct"` set and then we merge

File: test/bun.js/bun-server.test.ts [hunk 6]
```diff
@@ -26,4 +26,28 @@ describe("Server", () => {
...
+
+  test('abort signal on server', async ()=> {
+    {
+      let signalOnServer = false;
+      const server = Bun.serve({
+        async fetch(req) {
+          req.signal.addEventListener("abort", () => {
+            signalOnServer = true;
+          });
+          await Bun.sleep(3000);
```
Comment (Jarred-Sumner) [inline test/bun.js/bun-server.test.ts:38]: can we make these timeouts shorter so the test doesn't take as long, but still assert sequential order?

Comment (Jarred-Sumner) [reply test/bun.js/bun-server.test.ts:38]: after that i think its good to merge

**Pull request merged**